### ISSUE TITLE
Shared cache with dependencies

### ISSUE THREAD

**Issue by svbutko** [SDWebImage/SDWebImage#3360]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | iOS
 Platform Version        | 12.1 - 15.5
 SDWebImage Version      | 5.12.6
 Integration Method      | CocoaPods
 Xcode Version           | Xcode 13.4.1
 Repro rate              | all the time (100%)
 Repro with our demo prj | No
 Demo project link       | -

### Issue Description and Steps

Hi, I'm having an issue to create a single cache across the application and its dependencies:

I have an RN project which has 2 dependencies which in turn use SDWebImage.

The usage is exactly the same and primarily is focused around GIFs (SDAnimatedImageView) and setting a remote URL to a component (i.e. `sd_setImageWithURL`) with `SDWebImageProgressiveLoad` in options and `SDWebImageContextImageThumbnailPixelSize` in context, nothing specific apart from this.

The cached data ideally should be stored on disk.

The problem is following:
First dependency fetches a GIF, renders and caches it fine but the second one, supplied with the same URL, fetches it again and apparently caches it in a separate cache.

Is it possible to set the same cache for every dependency on project level or it has to be done through changes in dependencies? Or there's a different way of handling this issue?

Thank you.

**Comment by dreampiggy** [comment]
> supplied with the same URL, fetches it again and apparently caches it in a separate cache.

### Thumbnail Decoding

Whe using thumbnail decoding `SDWebImageContextImageThumbnailPixelSize`, actually by defaults, this will only store the **scaled down image** into memory cache, the original full data into disk cache.

So, it should be shared, from current design ?

### Transformer
Similiar but not always same as the thumbnail decoding. When using image transforminng, the transformed image will store to memory cache, and we will encode the transformed image again and store into disk cache. The original data is lost. To keep the original data into disk cache, you can control to use `SDWebImageContextOriginalStoreCacheType = .disk`. Read more in `SDWebImageContextOriginalStoreCacheType` and `SDWebImageContextOriginalQueryCacheType`

**Comment by dreampiggy** [comment]
Choose the seconds impl, PR will come.

**Comment by svbutko** [comment]
Checked it with [5.13.0](https://github.com/SDWebImage/SDWebImage/releases/tag/5.13.0), everything works, thank you a lot!

**Comment by heistings** [comment]
@dreampiggy Seems the feature has introduced a new issue. 

1. When I call with SDWebImageContextImageThumbnailPixelSize only once, I got two data disk cache. One is as large as the origin data, the other is smaller. This is pretty good.
2. But when I remove all the disk cache, and call no SDWebImageContextImageThumbnailPixelSize first, SDWebImageContextImageThumbnailPixelSize second, the origin data turns out to be much larger.(14KB above, but 34KB at this time)

**Comment by dreampiggy** [comment]
The second `SDWebImageContextImageThumbnailPixelSize(100, 100)` call will use the `query original data size from disk, then decode in memory, and re-encode to new image data`, this cause the issue.

**Comment by dreampiggy** [comment]
Is this bug exists on 5.13.2 ? I think I already add some check for the 4. Which should not store the new encoded data into disk if **it is came from the query original data size codepath**
